designs/node-repair.md

```diff
@@ -0,0 +1,93 @@
+
+### Option  1 (recommended) : Unhealthy condition set cloud provider interface  
+
+```
+type RepairPolicy struct {
+    // Type of unhealthy state that is found on the node
+    Type metav1.ConditionType 
+    // Status condition of when a node is unhealthy
+    Status metav1.ConditionStatus
+    // RepairAfter is the duration the controller will wait
+    // before attempting to terminate nodes that are marked for repair.
+    TolerationDuration time.Duration
+}
+
+type CloudProvider interface {
+  ...
+    // RepairPolicy is for CloudProviders to define a set Unhealthy condition for Karpenter 
+    // to monitor on the node. Customer will need 
+    RepairPolicy() []v1.RepairPolicy
+  ...
+}
+```
+
+The `RepairPolicy` will be a set of condition that the Karpenter controller will watch. The cloud provider can define compatibility with any node diagnostic agent, and track a list of node unhealthy condition types and a duration period to wait until a unhealthy state is considered a terminal: 
+
+1. A diagnostic agent, such as [node problem detector](https://github.com/kubernetes/node-problem-detector), will add a status condition on to a node 
+2. Karpenter will reconcile on nodes and match unhealthy node condition to cloud provider defined unhealthy condition 
+3. Unhealthy condition are added to the NodeClaim
+4. Disruption controller will forcefully terminate the the NodeClaim once the node has been in an unhealthy state for the duration specified by `repairAfter` of the unhealthy condition 
+
+### Option  2: cloud provider interface  for determining node health
+
+```
+type UnhealthyReason string
+
+type CloudProvider interface {
+  ...
+    // IsHealthy returns whether a NodeClaim is considered operating as expected
+    IsHealthy(context.Context, *v1.NodeClaim) (UnhealthyReason, error)
+  ...
+}
+```
+
+`IsHealthy` cloud provide methods allows each implementation of Karpenter to determine the healthy state of a NodeClaim and return a specified `UnhealthyReason` indicating that an action on the node is required. IsHealthy is a Turing complete operation and allows the cloud providers to determine when and on what condition to act on cleaning up the node. Cloud providers will also need to consider short lived failure prior to determining if a node is unhealthy. The `IsHealthy` follows the format defined by `IsDrifted`. 
+Main drawback of this option is providers will need to implement much more undifferentiated logic, without having tha ability to model duration and remediation on nodes. The reconciliation step for replacing an unhealthy node will look as such: 
```

Comment:
I'd like to point out that CloudProviders still have the ability to introduce delay -- it would just have to be hard-coded or defined in the NodeClass interface. You could just implement the delay in the IsHealthy() method